cdap-securestore-ext-cloudkms/src/main/java/co/cask/cdap/securestore/gcp/cloudkms/CloudKMSClient.java

```diff
@@ -0,0 +1,246 @@
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * Wrapper on {@link CloudKMS} client.
+ */
+class CloudKMSClient implements Closeable {
+  private static final Logger LOG = LoggerFactory.getLogger(CloudKMSClient.class);
+  // When created in the global location, Cloud KMS resources are available from zones spread around the world.
+  private static final String LOCATION_ID = "global";
+  // All the crypto keys are created under keyring named cdap
+  private static final String KEYRING_ID = "cdap";
+  private static final String ENCRYPT_DECRYPT = "ENCRYPT_DECRYPT";
+  private static final String CLOUD_KMS = "cloudkms";
+  private static final String PROJECT_ID = "projectId";
+  private static final String ACCOUNT_KEY = "accountKey";
+  private static final String METADATA_SERVER_API = "metadataApi";
+
+  private final CloudKMS cloudKMS;
+  private final String projectId;
+  private final String accountKey;
+  private final String metadataServerApi;
+  // In-memory cache to hold created crypto keys, this is to avoid checking if a given crypto key exists.
+  private final Set<String> crypoKeyList;
+
+  /**
+   * Constructs Cloud KMS client.
+   *
+   * @throws IOException if cloud kms client can not be created
+   */
+  CloudKMSClient(Map<String, String> properties) throws IOException {
+    this.projectId = properties.containsKey(PROJECT_ID) ? properties.get(PROJECT_ID) : getSystemProjectId();
+    this.accountKey = properties.getOrDefault(ACCOUNT_KEY, null);
+    this.metadataServerApi = properties.getOrDefault(METADATA_SERVER_API,
+                                                     "http://metadata.google.internal/computeMetadata/v1/project" +
+                                                       "/project-id");
+    this.crypoKeyList = new HashSet<>();
+    this.cloudKMS = createAuthorizedClient();
+  }
+
+  /**
+   * Get project id from the metadata server. Makes a request to the metadata server that lives on the VM,
+   * as described at https://cloud.google.com/compute/docs/storing-retrieving-metadata.
+   */
+  private String getSystemProjectId() {
+    try {
+      URL url = new URL(metadataServerApi);
+      HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+      try {
+        connection = (HttpURLConnection) url.openConnection();
+        connection.setRequestProperty("Metadata-Flavor", "Google");
+        connection.connect();
+        try (Reader reader = new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8)) {
+          return CharStreams.toString(reader);
+        }
+      } finally {
+        connection.disconnect();
+      }
+    } catch (IOException e) {
+      throw new RuntimeException("Unable to get project id from the environment. "
```

Comment:
Why throw runtime exception? The caller, which is the constructor has `throws IOException`.